### PR TITLE
Reduce disk-attach wait time in the Windows plugin (fixes #1780)

### LINKED ISSUE #1780: Skylight performance improvement - Investigate to reduce disk-attach wait time in the Windows plugin
The `fs.maxDiskAttachWaitSec` is set to 60 seconds to prevent failures during parallel volume requests, as access to the `powershell` utility is synchronized. This should be reduced once parallel disk identification is implemented. Currently, `fs.getDiskNum(..)` blocks during parallel execution due to synchronized access to `ps.Exec(..)`. Therefore, parallel volume creation in `e2e` tests block in `fs.DevAttachWait(..)` for a while and so we allow a long delay here.

See #1775 and [these lines](https://github.com/venilnoronha/docker-volume-vsphere/blob/0653f749f67c9e561e71964ff7cc7e133816ccbc/client_plugin/utils/fs/fs_windows.go#L34-L40) for more information.

/cc @shaominchen

### PR DESCRIPTION
**Pull request by shuklanirdesh82**
Reduce disk-attach wait time in the Windows plugin (fixes #1780)

Testbed scenario: Windows vm created on VMFS backed by ESX 6.0

Testing done: Yes 

**locally**:
1. Without change
```
--- PASS: Test (337.80s)
OK: 5 passed
PASS
ok      github.com/vmware/docker-volume-vsphere/tests/e2e       337.810s
```
2. With this change
```
--- PASS: Test (297.44s)
PASS
OK: 5 passed
ok      github.com/vmware/docker-volume-vsphere/tests/e2e       297.450s
```

**On CI**: https://ci.vmware.run/vmware/docker-volume-vsphere/1776
1. Master run: 
```
OK: 5 passed
--- PASS: Test (222.51s)
PASS
ok  	github.com/vmware/docker-volume-vsphere/tests/e2e	222.514s
```
2. with change: https://ci.vmware.run/vmware/docker-volume-vsphere/1777
```
OK: 5 passed
--- PASS: Test (208.74s)
PASS
```

test run logs: [winPluginE2ETests.txt](https://github.com/vmware/docker-volume-vsphere/files/1481020/winPluginE2ETests.txt)

/CC @shaominchen

### CODE REVIEW

File: client_plugin/utils/fs/fs_windows.go [hunk 1]
```diff
@@ -32,12 +32,7 @@ const (
 	FstypeDefault = ntfs
 
 	// maxDiskAttachWaitSec is the max time to wait for a disk to be attached.
-	// TODO: Reduce disk attach wait time once parallel disk identification is
-	// implemented. Currently, fs.getDiskNum(..) blocks during parallel execution
-	// due to synchronized access to ps.Exec(..). Therefore, parallel volume
-	// creation in e2e tests block in fs.DevAttachWait(..) for a while and so we
-	// allow a long delay here.
-	maxDiskAttachWaitSec = 120 * time.Second
+	maxDiskAttachWaitSec = 20 * time.Second
```
Comment (govint) [inline client_plugin/utils/fs/fs_windows.go]: Should also mention why its needed, even 20s seems like why its there? If each volume op is going to be in parallel there should be no wait time needed.

File: client_plugin/utils/powershell/powershell.go [hunk 2]
```diff
@@ -32,9 +32,6 @@ import (
 var (
 	// shell is the handle to a single open Powershell session.
 	shell ps.Shell
-
-	// mutex synchronizes access to the single PowerShell session.
```
Comment (govint) [inline client_plugin/utils/powershell/powershell.go]: How does this work? Can we say why the mutex was needed in the first place - is it like the powershll. Exec() function can handle only one call at a time. Is it to ensure that request and response are matched and one request (say volume create) doesn't get the response for a parallel volume Get(). This doesn't create a session per volume op. Its creating a session for the plugin instance. And all volume ops that are in flight will now use the single instance without any synchronization.

If we want a session per op then that session has to be created when a volume op request is received.

The init() function is run and the shell global is set when Go run time invokes this package init().

File: client_plugin/utils/fs/fs_windows.go (hunk 1 above)
Comment (shuklanirdesh82) [reply client_plugin/utils/fs/fs_windows.go]: It take 20-25 seconds to format `ntfs` type hence the wait time is required. I am attaching the volume creation log as shown below; please have a look at it. 

Setup detail: Volume creation from windows VM created on VMFS backed by ESX 6.0 

Command invocation:
```
PS C:\Program Files\Docker> date ; docker volume create -d vsphere Vol10 ; date                                             
Thu Nov 16 20:08:44 Coordinated Universal Time 2017                                                                             
vol10                                                                                                                       
Thu Nov 16 20:09:05 Coordinated Universal Time 2017 
```

vDVS log:
```
2017-11-16 20:08:45.1697236 +0000 GMT [ERROR] Failed to get volume meta-data error="Volume vol10 not found (file: /vmfs/volu
mes/datastore1/dockvols/11111111-1111-1111-1111-111111111111/vol10.vmdk)" name=vol10                                    
2017-11-16 20:08:46.989825 +0000 GMT [INFO] Attaching volume and creating filesystem name=vol10 fstype=ntfs                 
2017-11-16 20:08:46.989825 +0000 GMT [INFO] Waiting for a device change event                                                   
2017-11-16 20:08:48.4389095 +0000 GMT [INFO] Waiting for a watcher event volDev={3 224 17.0}                                    
2017-11-16 20:08:48.9668684 +0000 GMT [INFO] Watcher script executed stdout="EventTimedOut\n"                                   
2017-11-16 20:08:48.9668684 +0000 GMT [INFO] Successfully emitted event event=EventTimedOut                                     
2017-11-16 20:08:48.9668684 +0000 GMT [INFO] Waiting for a device change event                                                  
2017-11-16 20:08:48.9678642 +0000 GMT [INFO] Watcher emitted an event volDev={3 224 17.0} event=EventTimedOut                   
2017-11-16 20:08:50.7859716 +0000 GMT [INFO] Watcher script executed stdout="EventTimedOut\n"                                   
2017-11-16 20:08:50.7859716 +0000 GMT [WARNING] Couldn't emit event, continuing.. event=EventTimedOut                           
2017-11-16 20:08:50.7859716 +0000 GMT [INFO] Waiting for a device change event                                                  
2017-11-16 20:08:52.5390703 +0000 GMT [INFO] Watcher script executed stdout="EventTimedOut\n"                                   
2017-11-16 20:08:52.5390703 +0000 GMT [WARNING] Couldn't emit event, continuing.. event=EventTimedOut                           
2017-11-16 20:08:52.5390703 +0000 GMT [INFO] Waiting for a device change event                                                  
2017-11-16 20:08:53.4571233 +0000 GMT [INFO] Disk mapping script executed stdout="\n4\r\n" volDev={3 224 17.0}                  
2017-11-16 20:08:53.4571233 +0000 GMT [INFO] Successfully located disk volDev={3 224 17.0} diskNum=4                            
2017-11-16 20:08:53.4571233 +0000 GMT [INFO] Successfully mapped volDev to diskNum volDev={3 224 17.0} diskNum=4                
2017-11-16 20:08:53.4571233 +0000 GMT [INFO] Initiated watcher termination                                                      
2017-11-16 20:08:54.3011716 +0000 GMT [INFO] Watcher script executed stdout="EventTimedOut\n"                                   
2017-11-16 20:08:54.3011716 +0000 GMT [WARNING] Couldn't emit event, continuing.. event=EventTimedOut                           
2017-11-16 20:08:54.3011716 +0000 GMT [INFO] Watcher channels closed                                                            
2017-11-16 20:08:57.7903708 +0000 GMT [INFO] Disk mapping script executed volDev={3 224 17.0} stdout="\n4\r\n"                  
2017-11-16 20:08:57.7903708 +0000 GMT [INFO] Successfully located disk volDev={3 224 17.0} diskNum=4                            
2017-11-16 20:09:04.585771 +0000 GMT [INFO] Format disk script executed successfully fstype=ntfs label=vol10 volDev={3 224 1
7.0} diskNum=4 stdout="\r\nDriveLetter FileSystemLabel FileSystem DriveType HealthStatus OperationalStatus\r\n----------- ------
--------- ---------- --------- ------------ -----------------\r\n            vol10       NTFS       Fixed     Healthy      O
K               \r\n\r\n\r\n"                                                                                                   
2017-11-16 20:09:05.4068665 +0000 GMT [INFO] Volume and filesystem created fstype=ntfs name=vol10     
```

File: client_plugin/utils/powershell/powershell.go (hunk 2 above)
Comment (shuklanirdesh82) [reply client_plugin/utils/powershell/powershell.go]: As per below information, I guess locking mechanism was placed and ps.Exec() is getting blocked as described at https://github.com/vmware/docker-volume-vsphere/issues/1780#issue-251296227 hence the `120 sec` of threshold was added from https://github.com/vmware/docker-volume-vsphere/pull/1775#discussion_r134000743.

I have removed the dead code as well supplied the correct code, please review it again.

`
"Note that a single shell instance is not safe for concurrent use, as are remote sessions. You can have as many remote sessions using the same shell as you like, but you must execute commands serially. If you need concurrency, you can just spawn multiple PowerShell processes (i.e. call .New() multiple times)."
`

Reference: https://github.com/gorillalabs/go-powershell/blob/master/README.md

**Comment by shuklanirdesh82** [comment]
@govint Thanks for your comment!

The initial implementation was having some dead code that makes things complicated and hard to understand. I have corrected the code and tested on my local. 

We do have some tests which created volume parallel and they are working fine. Please give another look and share your feedback.

**Comment by shaominchen** [comment]
@shuklanirdesh82 Can you please provide a test result that shows the performance improvement? For example, the total e2e testing time before and after this improvement?

Moreover, I believe there must be some reason that we have introduced a single PowerShell session (which was removed in the current PR). I can't recall the details though - maybe @pshahzeb can comment on this?

**Comment by shuklanirdesh82** [comment]
@shaominchen I've updated the bug description for timing that you asked.

There is not much significant difference but the good thing is that threshold value `maxDiskAttachWaitSec` is set correctly `120 secs` to `30 secs`. 

Please refer https://github.com/vmware/docker-volume-vsphere/pull/1991#discussion_r151551849 for some analysis as well.

**Comment by shaominchen** [comment]
Based on the reference on https://github.com/gorillalabs/go-powershell/blob/master/README.md, it looks like spawning multiple PowerShell processes is a valid option to support concurrency. So this change makes sense to me.

We do need a test to measure the performance improvement - simply create a large number of volumes (say 100, or 500) in parallel and we should see the difference. It's very easy to add this test: almost same as existing TestValidName, but just creating more volumes.

File: tests/e2e/parallelvolumecreate_test.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+	wg.Wait()
+}
+
+func (s *ParallelVolumeCreateTestSuite) accessCheck(hostIP string, volList []string, c *C) {
+	isAvailable := verification.CheckVolumeListAvailability(hostIP, volList)
+	c.Assert(isAvailable, Equals, true, Commentf("Volume %s is not available after creation", volList))
+}
+
+// Create volume parallely
+func (s *ParallelVolumeCreateTestSuite) TestVolumeCreationParallel(c *C) {
```
Comment (shaominchen) [inline tests/e2e/parallelvolumecreate_test.go]: Why don't we simply add this new test into existing volumecreate_win_test.go? Most of the logic in this file is duplicate code.

Comment (shuklanirdesh82) [reply tests/e2e/parallelvolumecreate_test.go]: `volumecreate_win_test.go` hosts only different values like valid/invalid volume names.

Comment (shaominchen) [reply tests/e2e/parallelvolumecreate_test.go]: Yes, but if we add a new test to that file, this new test should be executed on windows only, right?

**Comment by shuklanirdesh82** [comment]
@shaominchen newly added test is successfully creating 10 disks consistently.

```
go test -v -timeout 50m -tags "runoncewin1 winutil" github.com/vmware/docker-volume-vsphere/tests/e2e
=== RUN   Test
2017/11/18 01:06:36 VM name is: windows
2017/11/18 01:06:36 START: ParallelVolumeCreateTestSuite.TestVolumeCreationParallel
2017/11/18 01:06:36 Creating volume [pvol_9_volume_990730] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_6_volume_356473] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_4_volume_113544] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_2_volume_280360] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_3_volume_416963] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_1_volume_416963] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_5_volume_388319] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_0_volume_388319] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_7_volume_1079825] on VM [10.20.104.126]
2017/11/18 01:06:36 Creating volume [pvol_8_volume_506529] on VM [10.20.104.126]
2017/11/18 01:08:45 Checking volume [pvol_6_volume_356473 pvol_3_volume_416963 pvol_2_volume_280360 pvol_1_volume_416963 pvol_4_volume_113544 pvol_9_volume_990730 pvol_5_volume_388319 pvol_7_volume_1079825 pvol_8_volume_506529 pvol_0_volume_388319] availability from VM [10.20.104.126]
2017/11/18 01:08:46 END: ParallelVolumeCreateTestSuite.TestVolumeCreationParallel
OK: 1 passed
--- PASS: Test (130.23s)
PASS
ok      github.com/vmware/docker-volume-vsphere/tests/e2e       130.238s
```

File: tests/e2e/parallelvolumecreate_test.go (hunk 3 above)
Comment (shuklanirdesh82) [reply tests/e2e/parallelvolumecreate_test.go]: The same amount of code has to be copied over.

**Comment by shaominchen** [comment]
@shuklanirdesh82 What about 20, 30, 40, 50? If it fails on 20 or 30, that's fine, but it will be useful to figure out how many concurrent requests we support.

**Pull request merged**